### PR TITLE
D3ASIM-3319: Refactored the Commands Enum and its usages

### PR DESCRIPTION
**Pull request by Hassan754**

### CODE REVIEW

**Review by faizan2590 — COMMENTED** [review]
Please update one of the existing setup client script([potentially this one](https://github.com/gridsingularity/d3a-api-client/blob/master/d3a_api_client/setups/test_redis_aggregator.py)) with examples targeting this feature.

**Comment by Hassan754** [comment]
> Please update one of the existing setup client script([potentially this one](https://github.com/gridsingularity/d3a-api-client/blob/master/d3a_api_client/setups/test_redis_aggregator.py)) with examples targeting this feature.

@faizan2590 I think you meant to comment on [this PR](https://github.com/gridsingularity/d3a-api-client/pull/109)

**Comment by faizan2590** [comment]
> > Please update one of the existing setup client script([potentially this one](https://github.com/gridsingularity/d3a-api-client/blob/master/d3a_api_client/setups/test_redis_aggregator.py)) with examples targeting this feature.
> 
> @faizan2590 I think you meant to comment on [this PR](https://github.com/gridsingularity/d3a-api-client/pull/109)

No, I meant this PR where you added [commands to enum](https://github.com/gridsingularity/d3a-api-client/pull/108/files#diff-93c7a331b07f8f0dcbc7528f6aab51e85ee91308f25deb93d1e09624d66867caR12-R16) but I don't see it being used an  examples in any setup files

**Comment by Hassan754** [comment]
> > > Please update one of the existing setup client script([potentially this one](https://github.com/gridsingularity/d3a-api-client/blob/master/d3a_api_client/setups/test_redis_aggregator.py)) with examples targeting this feature.
> > 
> > 
> > @faizan2590 I think you meant to comment on [this PR](https://github.com/gridsingularity/d3a-api-client/pull/109)
> 
> No, I meant this PR where you added [commands to enum](https://github.com/gridsingularity/d3a-api-client/pull/108/files#diff-93c7a331b07f8f0dcbc7528f6aab51e85ee91308f25deb93d1e09624d66867caR12-R16) but I don't see it being used as an examples in any setup files

@faizan2590 I only refactored an existing enum , I don't see the point of testing this , it's not even a feature

**Comment by faizan2590** [comment]
> > > > Please update one of the existing setup client script([potentially this one](https://github.com/gridsingularity/d3a-api-client/blob/master/d3a_api_client/setups/test_redis_aggregator.py)) with examples targeting this feature.
> > > 
> > > 
> > > @faizan2590 I think you meant to comment on [this PR](https://github.com/gridsingularity/d3a-api-client/pull/109)
> > 
> > 
> > No, I meant this PR where you added [commands to enum](https://github.com/gridsingularity/d3a-api-client/pull/108/files#diff-93c7a331b07f8f0dcbc7528f6aab51e85ee91308f25deb93d1e09624d66867caR12-R16) but I don't see it being used as an examples in any setup files
> 
> @faizan2590 I only refactored an existing enum , I don't see the point of testing this , it's not even a feature

In your old PR where you [abstracted batch commands API to python functions](https://github.com/gridsingularity/d3a-api-client/pull/80/files#diff-93c7a331b07f8f0dcbc7528f6aab51e85ee91308f25deb93d1e09624d66867caR4-R11), back then `update_offer/update_bid/device_info/grid_fees/market_stats /dso_market_stats ` weren't part of it and added in this PR. Right?

If yes, then simply while adding it now, you would have tested it locally too to validate these newly added commands are functional via any client script. So all Im asking is to either update any existing client script or create one using these as a reference for user

**Review by faizan2590 — COMMENTED** [review]
Thanks for adding integration test too, would be nice to add the newly added batch commands with at least a one liner explanation in README or as a reference in one of the example scripts

**Comment by Hassan754** [comment]
> Thanks for adding integration test too, would be nice to add the newly added batch commands with at least a one liner explanation in README or as a reference in one of the example scripts

@faizan2590 the only commands that should be added are update_offer / update_bid , as other stats are deprecated
I think I'll tag @fievelk to add the documentation for these two in his PR, as for now there is no `update_offer/update_bid` functions, they only exist in aggregator commads

File: d3a_api_client/enums.py
Comment (hannesdiedrich) [inline d3a_api_client/enums.py:31]: I thought we decided to deprecate these. I think this was only discussed in PRs and only mentioned in planning once. Shall I create a ticket for that @spyrostz ?

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [inline d3a_api_client/redis_client_base.py:50]: Can you please explain, why you dont want to subscribe to the response channels of the "new" commands e.g. UPDATE_BID ?

**Review by hannesdiedrich — COMMENTED** [review]
Two questions, will review again once addressed.

File: d3a_api_client/enums.py
Comment (Hassan754) [reply d3a_api_client/enums.py:31]: @hannesdiedrich Yes I am aware that we are to deprecate these , but this was not in the context of this ticket

File: d3a_api_client/redis_client_base.py
Comment (Hassan754) [reply d3a_api_client/redis_client_base.py:50]: Well it's not that I don't want to, I kept the existing callbacks and the main goal of the ticket was to remove the hardcoded strings from Commands class .
This is why I added the rest of commands to the enum and did not touch any logic in code
However, you are right ; I should probably add callbacks for the `update_offer`, `update_bid` , will do it

Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:50]: why not the responses of all of the `Commands`?

Comment (Hassan754) [reply d3a_api_client/redis_client_base.py:50]: I'm not sure why these showed now , why are they implemented from the d3a side and not in here? I I thought that them not being implemented is because they are not needed ?
However, I really think that this got way beyond this ticket, as we need to create callbacks for several responses , add new commands ( Both the Rest and Redis Devices/markets ) 
I suggest creating another ticket to deal with this, what do you think ?

Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:50]: You dont need to create new callbacks, all responses are handled in `_wait_and_consume_command_response`.

Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:50]: Or rather `_publish_and_wait`

Comment (Hassan754) [reply d3a_api_client/redis_client_base.py:50]: @hannesdiedrich yes but I mean the methods that would need to call this function, like `def update_offer` and so on

Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:50]: I see, you are right. But still I would vote for adding these in the frame of this story. @spyrostz What do you think?

Comment (Hassan754) [reply d3a_api_client/redis_client_base.py:50]: I'm in favor of not adding any response callback for the commands , as we have callbacks for events at this moment, and we have the on_event_or_response that can deal will all responses

File: d3a_api_client/enums.py
Comment (spyrostz) [reply d3a_api_client/enums.py:31]: Please create a ticket for that, these are probably not going to be needed again.

Comment (hannesdiedrich) [reply d3a_api_client/enums.py:31]: https://gridsingularity.atlassian.net/browse/D3ASIM-3351

File: d3a_api_client/redis_client_base.py
Comment (spyrostz) [reply d3a_api_client/redis_client_base.py:50]: I agree with Hannes here, the responses are needed, and you cannot get away easily with handling this on _on_event_or_response. These commands are actually the synchronous mode of operation of our client, where the execution is blocked until the command response from the d3a is received. Therefore they are very useful if a user wants to wait until one action has finished. 
You could potentially wait for everything that arrives on _on_event_or_response and then unblock only the response that matters to you, however I see a disadvantage compared to subscribing to the response. _on_event_or_response is on a different thread than the typical caller of these methods (on_market_cycle or on_tick) , therefore you would have to introduce locks / polling, whereas in this case this is already handled and no new code needs to be written.

Comment (Hassan754) [reply d3a_api_client/redis_client_base.py:50]: @spyrostz great, one more thing- do you think we should tackle the mentioned feature in this PR or you advise creating a ticket for it?

Comment (spyrostz) [reply d3a_api_client/redis_client_base.py:50]: @Hassan754 it is not in the context of the ticket I agree, however it should not be a big change, and this PR has not bloated yet, therefore it will be relatively easy to review. IMO let's deal with it here in order for the issue to not get deprioritized/forgotten. Thanks for dealing with it !

Comment (Hassan754) [reply d3a_api_client/redis_client_base.py:50]: Deal 👌

File: d3a_api_client/commands.py
Comment (hannesdiedrich) [inline d3a_api_client/commands.py]: I think this extra precaution could stay. Of course, it is only an extra safety-net for the programmer, but still. Thanks!

**Review by hannesdiedrich — APPROVED** [review]
One comment, once addressed, LGTM

**Review by hannesdiedrich — APPROVED** [review]
LGTM

File: d3a_api_client/rest_device.py
Comment (spyrostz) [inline d3a_api_client/rest_device.py:117]: I think that you can still leave it here, or am I mistaken ?

**Review by spyrostz — APPROVED** [review]
One comment, LGTM

File: d3a_api_client/rest_device.py
Comment (Hassan754) [inline d3a_api_client/rest_device.py:124]:
```suggestion
            response = self.dispatcher.wait_for_command_response('bid', transaction_id)
            log_bid_offer_confirmation(response)
```

**Review by hannesdiedrich — DISMISSED** [review]
LGTM

**Review by spyrostz — APPROVED** [review]
LGTM

**Review by faizan2590 — APPROVED** [review]
LGTM

**Pull request merged**
